File: LeetCode/valid_sudoku.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        // cout << "Check vertical" << endl;
        for(int i=0; i<9; i++){
            vector<bool> isAppeared(9, false);
            for(int j=0; j<9; j++){
                if(!isValid(board, i, j, isAppeared)){
                    return false;
                }
            }
        }
        // cout << "Check holizontal" << endl;
        for(int i=0; i<9; i++){
            vector<bool> isAppeared(9, false);
            for(int j=0; j<9; j++){
                if(!isValid(board, j, i, isAppeared)){
                    return false;
                }
            }
        }
        // cout << "Check block" << endl;
        for(int i=0; i<9; i++){
            vector<bool> isAppeared(9, false);
            for(int j=0; j<9; j++){
                if(!isValid(board, (i%3)*3 + j%3, (i/3)*3 + j/3, isAppeared)){
                    return false;
                }
            }
        }
        return true;
    }
private:
    bool isValid(vector<vector<char>>& board, int x, int y, vector<bool>& isAppeared){
        // cout << "x = " << x << ", y = " << y << endl;
        if(board[x][y] == '.'){
            return true;
        }
        int num = board[x][y] - '0';
        if(isAppeared[num]){
            // cout << "x = " << x << ", y = " << y << ", num = " << num << endl;
            return false;
        }
        isAppeared[num] = true;
        return true;
    }
};

int main(void){
    vector<vector<char>> input = {
        {'5','3','.','.','7','.','.','.','.'},
        {'6','.','.','1','9','5','.','.','.'},
        {'.','9','8','.','.','.','.','6','.'},
        {'8','.','.','.','6','.','.','.','3'},
        {'4','.','.','8','.','3','.','.','1'},
        {'7','.','.','.','2','.','.','.','6'},
        {'.','6','.','.','.','.','2','8','.'},
        {'.','.','.','4','1','9','.','.','5'},
        {'.','.','.','.','8','.','.','7','9'}
    };
    Solution sol;
    cout << (sol.isValidSudoku(input) ? "true" : "false") << endl;
    return 0;
}

File: LeetCode/valid_sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        return self.isValidRow(board) and self.isValidCol(board)\
            and self.isValidSubBox(board)
    
    def isValidRow(self, board):
        for c in range(len(board[0])):
            num_set = set()
            for r in range(len(board)):
                if board[r][c] == ".":
                    continue

                if board[r][c] in num_set:
                    return False
                num_set.add(board[r][c])
        return True
    
    def isValidCol(self, board):
        for r in range(len(board)):
            num_set = set()
            for c in range(len(board[0])):
                if board[r][c] == ".":
                    continue

                if board[r][c] in num_set:
                    return False
                num_set.add(board[r][c])
        return True

    def isValidSubBox(self, board):
        num_sets = [set() for _ in range(9)]
        for r in range(len(board)):
            for c in range(len(board[0])):
                if board[r][c] == ".":
                    continue

                i = (r // 3) * 3 + c // 3
                if board[r][c] in num_sets[i]:
                    return False
                num_sets[i].add(board[r][c])
        return True

if __name__ == "__main__":
    board1 = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]
    board2 = [
        ["8","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]
    sol = Solution()
    res = sol.isValidSudoku(board2)
    print(res)